src/librustc_mir/interpret/intern.rs

```diff
@@ -341,12 +340,19 @@ pub fn intern_const_alloc_recursive<M: CompileTimeMachine<'mir, 'tcx>>(
             // We can't call the `intern_shallow` method here, as its logic is tailored to safe
             // references and a `leftover_allocations` set (where we only have a todo-list here).
             // So we hand-roll the interning logic here again.
-            if base_intern_mode != InternMode::Static {
-                // If it's not a static, it *must* be immutable.
-                // We cannot have mutable memory inside a constant.
-                // FIXME: ideally we would assert that they already are immutable, to double-
-                // check our static checks.
-                alloc.mutability = Mutability::Immutable;
+            match base_intern_mode {
+                InternMode::Static => {},
+                InternMode::Const | InternMode::ConstBase => {
+                    // If it's not a static, it *must* be immutable.
+                    // We cannot have mutable memory inside a constant.
+                    // We use `delay_span_bug` here, because this can be reached in the presence
+                    // of fancy transmutes.
+                    if alloc.mutability == Mutability::Mutable {
+                        // For better errors later, mark the allocation as immutable
+                        alloc.mutability = Mutability::Immutable;
+                        ecx.tcx.sess.delay_span_bug(ecx.tcx.span, "mutable allocation in constant");
```

Comment:
Oh wow. Can we really get away with this?^^

Reply:
The previous hard ICE only triggered *after* we already reported a really helpful error. Repeating this error reporting here would just be noise imo

Reply:
Wait we already had a hard ICE for this? Where?

Reply:
Oh we didn't... This was an intermediate state of this PR, but that caused ICEs in case other errors were already reported.

Reply:
Ah. Okay so my remark remains that I am surprised that we can get away with a delayed ICE here. But I like it. :D

I just wonder if this will remain the case as CTFE becomes more powerful.

Reply:
Well... we'd go from code failing to compile with an error to code failing to compile with an ICE, so... it's not like it's fragile, just annoying and we'll address it then.